cnf-certification-test/accesscontrol/suite.go

```diff
@@ -398,60 +399,102 @@ func testPodServiceAccount(env *provider.TestEnvironment) {
-			logrus.Infof("%s has an empty serviceAccountName, skipping.", put.String())
+		if put.Pod.Spec.ServiceAccountName == "" {
+			logrus.Infof("%s has an empty or default serviceAccountName, skipping.", put.String())
 			continue
 		}
 
-		// Get any rolebindings that do not belong to the pod namespace.
-		roleBindings, err := rbac.GetRoleBindings(put.Namespace, put.Spec.ServiceAccountName)
-		if err != nil {
-			failedPods = append(failedPods, put.Name)
-		}
+		logrus.Infof("%s has a serviceAccountName: %s, checking role bindings.", put.String(), put.Spec.ServiceAccountName)
 
-		if len(roleBindings) > 0 {
-			logrus.Warnf("Pod: %s/%s has the following role bindings: %s", put.Namespace, put.Name, roleBindings)
-			tnf.ClaimFilePrintf("Pod: %s/%s has the following role bindings: %s", put.Namespace, put.Name, roleBindings)
-			failedPods = append(failedPods, put.Name)
+		podIsCompliant := true
+		// Loop through the rolebindings and check if they are from another namespace
+		for rbIndex := range env.RoleBindings {
+			// Short circuit if the role binding and the pod are in the same namespace.
+			if env.RoleBindings[rbIndex].Namespace == put.Namespace {
+				continue
+			}
+
+			// If we make it to this point, the role binding and the pod are in different namespaces.
+			// We must check if the pod's service account is in the role binding's subjects.
+
+			found := false
+			for _, subject := range env.RoleBindings[rbIndex].Subjects {
+				// If the subject is a service account and the service account is in the same namespace as the pod, then we have a failure
+				//nolint:gocritic
+				if subject.Kind == rbacv1.ServiceAccountKind && subject.Namespace == put.Namespace && subject.Name == put.Spec.ServiceAccountName {
+					failMsg := fmt.Sprintf("Pod: %s/%s has the following role bindings that do not live in the same namespace: %s", put.Namespace, put.Name, env.RoleBindings[rbIndex].Name)
+					logrus.Warnf(failMsg)
+					tnf.ClaimFilePrintf(failMsg)
+
+					// Add the pod to the non-compliant list
+					nonCompliantObjects = append(nonCompliantObjects,
+						testhelper.NewPodReportObject(put.Namespace, put.Name,
+							"Non-compliant because the role bindings used by this pod do not live in the same namespace", false).
+							AddField(testhelper.RoleBindingName, env.RoleBindings[rbIndex].Name).
+							AddField(testhelper.RoleBindingNamespace, env.RoleBindings[rbIndex].Namespace).
+							AddField(testhelper.ServiceAccountName, put.Spec.ServiceAccountName))
+					found = true
+					podIsCompliant = false
+					break
+				}
+			}
+			// Break of out the loop if we found a role binding that is out of namespace
+			if found {
+				break
+			}
+		}
+		// Add pod to the compliant object list
+		if podIsCompliant {
+			compliantObjects = append(compliantObjects,
+				testhelper.NewPodReportObject(put.Namespace, put.Name, "Compliant because the role bindings used by this pod (applied by the service account) live in the same namespace", true))
```

Comment:
minor/optional:
Compliant because the role bindings used by this pod (applied by the service account) live in the same namespace -> Compliant because **all** the role bindings used by this pod (applied by the service account**s**) live in the same namespace